### PR TITLE
Fixes #3 and #4

### LINKED ISSUE #3: Don't let errors pass silently
When row has different number of fields than described in namedtuple, this row is silently converted to None.

Not sure if this is a good thing to let errors pass it silently. I like Python Pandas approach to this:

> **error_bad_lines** : boolean, default True
> 
> Lines with too many fields (e.g. a csv line with too many commas) will by default cause an exception to be raised, and no DataFrame will be returned. If False, then these “bad lines” will dropped from the DataFrame that is returned. (Only valid with C parser)
> 
> http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html

### LINKED ISSUE #4: Forbidden final backslash?
It is possible to generate a TSV using `tsv.to`, containing backslash in a last field, but not possible to read this output using `tsv.un`.

Also, there is no explanation, why final backslash is forbidden?

### PR DESCRIPTION
**Pull request by sirex**
@solidsnack I just finished working on #3 and #4 and saw, that you have fixed #4.

My approach to backslash handling is a bit different, I check pairs of characters, where first one is backslash and second is escape character. Only `tnr\` characters can be used as escape characters. All other characters or no character at all (at the end of field) are threated as error.

Escape decoding is also integrated with error handling parameter, one can make errors pass silently (`error_bad_lines=False`), but by default it raises exception.

### CODE REVIEW

**Comment by solidsnack** [comment]
> My approach to backslash handling is a bit different, I check pairs of characters, where first one is backslash and second is escape character. Only tnr\ characters can be used as escape characters. All other characters or no character at all (at the end of field) are threated as error.

[However](http://dataprotocols.org/linear-tsv/):

> If a backslash precedes another character but does not form one of the escape sequences above, it is a “superfluous backslash” and is removed from the field on read.

File: tests.py [hunk 1]
```diff
@@ -13,18 +13,16 @@ def test_un():
         r'1	a',
         r'2	\t',
         r'3	\n',
-        r'4	\\',
+        r'4	\\\\',
```
Comment (solidsnack) [inline tests.py:17]: With `\\` in a raw string, this actually tested the right thing: that a final pair of backslashes was okay but a dangling final backslash was not. I am not sure that `\\\\` tests anything different but it does seem less clear.

File: tsv.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
 # coding: utf-8
 from __future__ import unicode_literals
 
+import re
 import six
+import warnings
 
 
-def un(source, row=list):
+def un(source, wrapper=list, error_bad_lines=True):
```
Comment (solidsnack) [inline tsv.py:9]: This seems like a good direction to go in.

File: tsv.py [hunk 3]
```diff
@@ -29,46 +31,89 @@ def un(source, row=list):
...
+        if check_line_consistency(columns, values, i, error_bad_lines):
+            yield wrapper(values)
 
-def parse_namedtuple(line, namedtuple):
-    try:
-        return namedtuple._make(parse_line(line))
-    except TypeError:
-        return None
+
+def is_namedtuple(obj):
```
Comment (solidsnack) [inline tsv.py:54]: Good to move this out/aside.

**Comment by solidsnack** [comment]
I think the high-level API changes you have made are basically right but I question the use of regexen and more generally, the field parser.

We could restore the field parser to the one in `solidsnack:master` and keep all your other changes if you think that actually will work end to end.

Sorry for the many changes in a short period of time.

**Comment by sirex** [comment]
I just restored previous behavior, that was implemented by you. But still it uses regex. To me regex approach looks more readable and more flexible, since I could completely change backslash interpretation by changing two-three lines. Not sure about performance.

File: tests.py (hunk 1 above)
Comment (sirex) [inline tests.py:17]: Added that case too: https://github.com/solidsnack/tsv/pull/6/commits/0fda23221a1a7d2e88690cbdfc0648b2f7ebafbe

**Comment by solidsnack** [comment]
Flexibility is not necessarily a benefit for something implementing a spec like this. In truth I am concerned about both efficiency and clarity.

**Comment by solidsnack** [comment]
Your changes have been released to PyPI (as part of https://github.com/solidsnack/tsv/commit/9fd239975765dff66f2341aeb9cee266085aa5fb): https://pypi.python.org/pypi/linear-tsv/0.99.1

**Comment by sirex** [comment]
Thanks!

**Comment by sirex** [comment]
Just in case, decided to measure performance of `partition` vs `re`, here is the code that was used:

``` python
import re

ESCAPE_DECODE_RE = re.compile(r'\\(.)|\\()$', flags=(re.MULTILINE | re.DOTALL))
ESCAPE_CHARS = {
    't': '\t',
    'n': '\n',
    'r': '\r',
    '\\': '\\',
}


def parse_field_partition(s):
    o = ''
    if s == '\\N':
        return None
    before, sep, after = s.partition('\\')
    while sep != '':
        o += before
        if after == '':
            raise ValueError('Final backslash is forbidden.')
        if after[0] in ESCAPE_CHARS:
            o += ESCAPE_CHARS[after[0]]
            before, sep, after = after[1:].partition('\\')
        else:
            before, sep, after = after.partition('\\')
    else:
        o += before
        return o


def _escape_decode(match):
    char = match.group(1)
    return ESCAPE_CHARS.get(char, char or '')


def parse_field_re(s):
    if s == '\\N':
        return None
    return ESCAPE_DECODE_RE.sub(_escape_decode, s)


if __name__ == '__main__':
    from timeit import timeit

    def run(s):
        ns = dict(globals(), s=s)
        a = timeit("parse_field_partition(s)", globals=ns)
        b = timeit("parse_field_re(s)", globals=ns)
        p = 100 / max(a, b) * (max(a, b) - min(a, b))
        s = '%s...(+%d chars)' % (s[:16], len(s) - 16) if len(s) > 24 else s
        print('%9.2f | %5.2f | %7.0f%% | %r' % (a, b, p, s))

    print('partition | re    | % faster | string')
    print('----------+-------+----------+---------')

    run('1\t2')
    run('\t'.join(map(str, range(100))))
    run('\t'.join(map(str, range(1000))))
```

And these are results:

```
partition | re    | % faster | string
----------+-------+----------+---------
     0.43 |  0.74 |      41% | '1\t2'
     0.45 |  0.95 |      53% | '0\t1\t2\t3\t4\t5\t6\t7\t...(+273 chars)'
     0.56 |  3.44 |      84% | '0\t1\t2\t3\t4\t5\t6\t7\t...(+3873 chars)'
```

`partition` is at least two times faster than `re`.

**Comment by solidsnack** [comment]
It seems reasonable on the face of it -- it has less work to do.

It is also makes clear and precise how the matching would be performed in something like C or Lua, which do not have native regexen.

**Pull request merged**